### ISSUE TITLE
[BUG][wrong-code] A shared `T.gemm` operand sliced along MN silently reads the wrong K panel when `block_K` spans more than one swizzle atom

### ISSUE THREAD

**Issue by bigSheep123** [tile-ai/tilelang#2964]
## Summary

Passing a shared buffer **sliced along MN** as a `T.gemm` operand silently computes the wrong result whenever `block_K` spans more than one swizzle atom. No warning, no assertion, no illegal access — the kernel reads real data from the wrong place, so it just returns wrong numbers.

The same GEMM written with one shared buffer per slice is correct, which makes the slice/no-slice pair a clean A/B.

## Reproduction

`main@6af825e0`, H200 (sm_90a), CUDA 13.2, torch 2.13.

```python
import torch

import tilelang
import tilelang.language as T

M, N, K = 64, 128, 128
BM, BN, BK, W0 = 64, 128, 128, 64  # split BN into W0 + (BN - W0) = 64 + 64


def build(sliced: bool):
    @T.prim_func
    def main(
        a: T.Tensor((M, K), "bfloat16"),
        b: T.Tensor((N, K), "bfloat16"),
        c: T.Tensor((M, N), "bfloat16"),
    ):
        with T.Kernel(T.ceildiv(N, BN), T.ceildiv(M, BM), threads=128) as (bx, by):
            a_s = T.alloc_shared((BM, BK), "bfloat16")
            c0 = T.alloc_fragment((BM, W0), "float32")
            c1 = T.alloc_fragment((BM, BN - W0), "float32")
            T.clear(c0)
            T.clear(c1)
            if sliced:
                # One shared buffer, consumed as two row slices.
                b_s = T.alloc_shared((BN, BK), "bfloat16")
            else:
                # Same math, one buffer per slice.
                b_s0 = T.alloc_shared((W0, BK), "bfloat16")
                b_s1 = T.alloc_shared((BN - W0, BK), "bfloat16")
            for ko in T.Pipelined(T.ceildiv(K, BK), num_stages=0):
                T.copy(a[by * BM, ko * BK], a_s)
                if sliced:
                    T.copy(b[bx * BN, ko * BK], b_s)
                    T.gemm(a_s, b_s[0:W0, :], c0, transpose_B=True)
                    T.gemm(a_s, b_s[W0:BN, :], c1, transpose_B=True)
                else:
                    T.copy(b[bx * BN, ko * BK], b_s0)
                    T.copy(b[bx * BN + W0, ko * BK], b_s1)
                    T.gemm(a_s, b_s0, c0, transpose_B=True)
                    T.gemm(a_s, b_s1, c1, transpose_B=True)
            T.copy(c0, c[by * BM, bx * BN])
            T.copy(c1, c[by * BM, bx * BN + W0])

    return tilelang.compile(main, out_idx=[2], target="cuda")


torch.manual_seed(0)
a = torch.randn(M, K, device="cuda", dtype=torch.bfloat16)
b = torch.randn(N, K, device="cuda", dtype=torch.bfloat16)
ref = a.float() @ b.float().T

for sliced in (False, True):
    err = (build(sliced)(a, b).float() - ref).abs().max().item()
    print(f"{'one buffer, two slices' if sliced else 'two buffers      ':22s}  max_err = {err:.4f}")
```

```
two buffers             max_err = 0.1248     <- bf16 accumulation noise
one buffer, two slices  max_err = 43.3982    <- wrong
```

## Scope

Both operands and both slice axes, `block_M = block_N = 128`, slicing 64 + 64:

| operand | `block_K=64` | `block_K=128` | `block_K=256` |
|---|---|---|---|
| K-major A (`A_s[0:64, :]`) | ok | **err ≈ 113** | **err ≈ 121** |
| K-major B (`B_s[0:64, :]`, `transpose_B=True`) | ok | **err ≈ 101** | **err ≈ 107** |
| MN-major A (`A_s[:, 0:64]`, `transpose_A=True`) | ok | ok | ok |
| MN-major B (`B_s[:, 0:64]`) | ok | ok | ok |

Independent of `threads`, `num_stages`, warp specialization, and dtype (fp16 and bf16 both). MN-major operands are unaffected — that branch of the offset formula does not use the term below.

`block_K = 64` is correct only by accident: at bf16 that is exactly one 128B swizzle atom, so the buggy term is multiplied by zero (see below).

## Root cause

`tilelang/cuda/intrinsics/macro/wgmma_macro_generator.py`, `wgmma_ss_atom` / `wgmma_rs_atom`:

```python
B_offset = (ki // bk_atom_size) * n_dim * b_swizzle_atom_elems + ...
```

When `block_K` exceeds one swizzle atom, the shared layout splits the K axis into panels and stores them **panel-major**. A `(128, 128)` bf16 tile decodes as

```
Sw<3,3,3> o 0 o (128,(64,2)):(64,(1,8192))
                              ^^^^ panel stride = 128 rows x 64 elems
```

so advancing to the next K panel steps by **the buffer's** MN extent. The formula instead reconstructs that step from `n_dim`, the extent of *this operand*. The two coincide only when the operand covers the whole buffer; for a slice `n_dim` shrinks while the panel spacing does not.

Concretely, for the repro above (bf16, 64-element atom, `bk_atom_size = 4`, so `ki // 4` selects the panel):

| operand | slice origin (correct) | `ki >= 4` should read | actually reads |
|---|---|---|---|
| `b_s[0:64, :]` | elem 0 | 8192 = row 0, panel 1 | 4096 = **row 64, panel 0** |
| `b_s[64:128, :]` | 4096 ✓ | 12288 = row 64, panel 1 | 8192 = **row 0, panel 1** |

Emitted CUDA — the panel step is 8192 bytes where it should be 16384:

```cpp
tl::initialize_wgmma_descriptor<1, 1, 64>(desc_b_1, (char*)&b_s[0] + 8192);  // origin: correct
tl::wgmma_ss<...>(..., desc_b_1 + (((ki >> 2) * 8192 + (ki & 3) * 32) >> 4), ...);
//                                            ^^^^ should be 16384
```

Two consequences worth calling out:

- The **slice origin is correct** (`slice_byte_offset` is derived from the layout). Only the panel step is wrong, so the first K atom of every operand is right and everything past it is wrong — the error is a partial-garbage mix rather than an obvious blow-up.
- The two slices read **each other's rows**, all within the buffer, so there is no out-of-bounds signal.

The same expression is duplicated in five more places, all with the same defect:

- `wgmma_sp_macro_generator.py` — sparse Hopper, `wgmma_ss` and `wgmma_rs`
- `tcgen05_macro_generator.py` — Blackwell, `tcgen05_ss_atom`, `tcgen05_ts_atom`, `tcgen05_blockscaled_atom`

## Why nothing caught it

`tilelang/language/gemm_op.py` used to assert `A_offset[-2] == 0`, i.e. reject an operand with a non-zero offset on the second-to-last axis. That assertion was dropped when the region-based lowering landed, on the premise that "every lowering consumes the complete operand BufferRegions". That is half true: the *origin* is consumed from the region correctly, but the *panel stride* is still reconstructed from emitter-level tile dims. The guard was removed and no equivalent check replaced it.

Existing coverage misses it on both backends:

- `test_tilelang_kernel_gemm_with_stride.py` does slice a shared operand, but at `block_K = 32` (single atom) and on the pre-Hopper `mma` path.
- `test_tilelang_language_tcgen05_sliced_operands.py` slices only along **K**, which is exactly the axis where the panel term is unused. MN slicing is validated for atom alignment but never numerically.

## Environment

- tilelang `main@6af825e0`
- H200 (sm_90a), nvcc 13.2, driver 595.71.05, torch 2.13.0+cu130
- Also reproduces with `TL_DISABLE_WARP_SPECIALIZED` both on and off

I have a fix and will open a PR shortly.
